### PR TITLE
Enable editing of controls inside viewports in editor

### LINKED ISSUE #20619: Can't move node inside Viewport
**Godot version:**
3.0/3.1-dev

**Issue description:**
When you place node (e.g. Sprite) inside a Viewport, it can't be moved in the editor - you need to manually change position in inspector. While with select tool this is understandable behavior, because your viewport display node (ViewportContainer in my case) is selected instead of sprite, move tool doesn't work either.

**Steps to reproduce:**
1. Add a ViewportContainer (or any Control node)
2. Add Viewport as a child
3. Add Sprite to the Viewport and make sure it displays
4. Press W to select move tool
5. Try to move the Sprite
6. Spoiler: it won't move

**Minimal reproduction project:**
Contains aforementioned scene setup.
[TestProject.zip](https://github.com/godotengine/godot/files/2243026/TestProject.zip)

### LINKED ISSUE #60088: Layout preset buttons not showing for Window / Popup nodes and their children Control nodes
### Godot version

v4.0.alpha6.official [e4f0fc50f]

### System information

Linux Pop!_OS 21.10, Nvidia GT710, Nvida X Server Settings 470.75.01

### Issue description

When we create a Window or a Popup node and we select them in the scene tree in the editor, they don't get the rectangle highlight and red circles to used to resize the nodes. The same happens to any of their children that inherit Control (probably other types too), and the buttons with the presets for anchors and sizing don't appear either, making it almost impossible to layout the GUI inside Window or Popup nodes.  It is possible to layout the GUI via the Inspector tab changing the property `layout_mode` to Anchors (when possible), but that's not very convenient.
The theme for those nodes don't seem to be working properly either.

### Steps to reproduce

Add a Window or any Popup node to the scene tree;
Add a Control as a child and some other node that inherits control as its child
Try to layout the GUI.

### Minimal reproduction project

[popup_test.zip](https://github.com/godotengine/godot/files/8457963/popup_test.zip)

### LINKED ISSUE #79289: Can no longer select contents of a WindowDialog using mouse (In Editor)
### Godot version

4.1

### System information

Linux Mint 21.1 Cinnamon, Graphics: Intel Corporation HD Graphics 620, Processor: Intel© Core™ i5-7200U CPU @ 2.50GHz × 2

### Unable to select anything using mouse inside "Window" and "SubViewport" nodes **In Editor**
In Godot 3 i was able to:

1. Select items/nodes IN the WindowDialog (and other popup nodes) **"Using the mouse"**
2. There was an **Orange-Yellowish boundary** that i used to "Visualize" the extents of the contents
Both of these were very useful to me :sob:  (I know i can do the same by changing values in properties but visualizing it is important too)

https://github.com/godotengine/godot/assets/77773850/61909bcf-9dc1-42dc-97dd-01e032bb7e58

And here's What happens in godot 4 (NO visual selection rectangle appearing in 2d viewport at all):

https://github.com/godotengine/godot/assets/77773850/5be91939-3eeb-4dc1-933e-955b071a3a13

I think this is due to Window now being child of Viewport
(In Viewport (for Godot 3) one was ALSO unable to do the above 2 things so it makes sense that Godot 4 "Window" node would do the same)

If it is possible please allow these 2 features (og Godot 3) to continue existing in Godot 4

### Steps to reproduce

(this doesn't require those)

### Minimal reproduction project

(not needed)

### PR DESCRIPTION
**Pull request by Black-Cat**
Fixes #79289

I've tested:
* Selection
* Box Selection
* Anchors
* Moving
* Rotating
* Scaling
* Pivot
* Temporary pivot
* Smart Snapping
* Grid Snapping
* Helpers
* Focus

*Bugsquad edit:*
Fixes #20619
Fixes #60088

### CODE REVIEW

File: editor/plugins/canvas_item_editor_plugin.cpp [hunk 1]
```diff
@@ -610,23 +610,29 @@ void CanvasItemEditor::_find_canvas_items_at_pos(const Point2 &p_pos, Node *p_no
...
 		return;
 	}
-	if (Object::cast_to<Viewport>(p_node)) {
-		return;
-	}
 
 	const real_t grab_distance = EDITOR_GET("editors/polygon_editor/point_grab_radius");
 	CanvasItem *ci = Object::cast_to<CanvasItem>(p_node);
 
+	Transform2D p_xform = p_canvas_xform;
```
Comment (KoBeWi) [inline editor/plugins/canvas_item_editor_plugin.cpp:617]:
```suggestion
	Transform2D xform = p_canvas_xform;
```
`p_` prefix is reserved for parameters.

**Review by KoBeWi — APPROVED** [review]
Tested various 2D editor interactions and didn't see any obvious side-effects. It works surprisingly well.

File: editor/plugins/canvas_item_editor_plugin.cpp (hunk 1 above)
Comment (Black-Cat) [reply editor/plugins/canvas_item_editor_plugin.cpp]: updated

**Comment by kitbdev** [comment]
Some nodes have an incorrect gizmo position when in a Viewport not at the origin, like position indicators for Node2Ds or handles for Line2D, CollisionShape2D, etc.

image

image

image

Selecting works and using the handles works, they're just offset and not rotated or scaled.

Also, If the viewport gets rotated or scaled, using the move or scale tools aren't correct, though the resize gizmos are fine.

image

**Review by kitbdev — APPROVED** [review]
The issues I mentioned before are all fixed. Looks good!

**Review by Calinou — COMMENTED** [review]
Tested locally (rebased on top of `master` https://github.com/godotengine/godot/commit/0eadbdb5d0709e4e557e52377fa075d3e2f0ad1f), it mostly works as expected.

On [GUI in 3D](https://github.com/godotengine/godot-demo-projects/tree/master/viewport/gui_in_3d), I can see the editing controls and interact with them, but nothing appears in the 2D editor viewport:

image

All nodes are set to be visible in the 2D scene tree.

I can see the preview in the inspector though:

image

**Comment by kitbdev** [comment]
> On [GUI in 3D](https://github.com/godotengine/godot-demo-projects/tree/master/viewport/gui_in_3d), I can see the editing controls and interact with them, but nothing appears in the 2D editor viewport:

They also do not appear in 4.3 and in master. This is intentional since [SubViewport](https://docs.godotengine.org/en/latest/classes/class_subviewport.html) do not draw anything by themselves. In that scene, the SubViewport is rendered to a MeshInstance3D, so it will only show in the 3D Editor.

So instead, the nodes in a SubViewport without a SubViewportContainer should not be selectable or editable in the canvas editor.

**Comment by akien-mga** [comment]
Thanks!

**Comment by KoBeWi** [comment]
Unintended consequence is that you can move nodes outside current scene, which is probably what the original check wanted to prevent:

https://github.com/user-attachments/assets/d5451ebe-df40-49cb-a763-dfe3a539200b

Granted, you need a custom EditorPlugin to select such nodes. We could use `in_part_of_edited_scene()` to prevent this, but it's not really a problem I think 🤔

**Pull request merged**